src/app/services/battle.service.ts

```diff
@@ -0,0 +1,81 @@
+import { Injectable } from '@angular/core';
+import { Pokemon } from './../models/pokemon.interface';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class BattleService {
+  public counterArr: any = [];
+
+  constructor() { }
+
+  public isAlive(pokemon: Pokemon): boolean {
+    return pokemon.health > 0;
+  }
+
+  public basePunch(a: Pokemon, b: Pokemon): number {
+    let currentAbility: number;
+    //let result: number;
+    if (this.counterArr.indexOf(a.state) !== -1) {
+      this.counterArr.push(a.state);
+    }
+    currentAbility = this.getValOfObj(a.ability);
+    const random: number = this.generateRandom(currentAbility);
+    // result = b.health - random;
+
+    return random;
+  }
+
+  public specAttack(a: Pokemon, b: Pokemon): number {
+    if (this.isSpecAttack(a)) {
+      this.counterArr = this.counterArr.filter((el: string) => {
+        return el !== a.state;
+      });
+      this.counterArr.push(a.state);
+      if (a.specAttack.type === 'damage') {
+        return b.health - this.generateRandom(a.specAttack.points);
```

Comment:
you can have some kind of a map where you say what you would like to do with heath. Example:
```
enum attackType {
  recovery = 'recovery',
  defence = 'defence',
  damage = 'damage',
}

const attackMap: Map<attackType, number> = new Map([
  [attackType.recovery, 1],
  [attackType.defence, 1],
  [attackType.damage, -1],
]);

```
then in your function you could simply have one line instead of three if else:
`return attackMap.get(a.specAttack.type) *  a.health + this.generateRandom(a.specAttack.points);`